### PR TITLE
Writing flow: Copy whole block if no text is selected

### PR DESCRIPTION
**Pull request by mcsf**
## Description

* Allow a single block to be copied or cut as a whole via the browser's _copy_ and _cut_ actions.

That is, when a single block is selected but no text is selected, copying (such as using <kbd>CTRL-C</kbd>) and cutting (<kbd>CTRL-X</kbd>) now grab the entire block. This is an extension of the behaviour that was in place for selections of multiple blocks.

Subsequently added: 
* Flash block(s) when copied.
* Show snackbar notifications for copy and cut actions, whether one or many blocks are copied.

copy-flash-and-cut-notifications

## Questions

I extended this to _cut_ (<kbd>CTRL-X</kbd>) to see how it feels. Does it make sense?

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by ellatrix** [comment]
Sounds good. The only thing I would find a bit strange is that there's no visual indication of what you're copying. If you select some text, it's visible, and clear what you copy. If you select some blocks, you see the selection and what you copy. That's not the case there. On the other hand the shortcut doesn't do anything, so it doesn't really hurt to add it. :)

**Comment by ellatrix** [comment]
Do you think an e2e test could be useful? The shortcuts work in content editable fields in our e2e tests.

**Review by ellatrix — APPROVED** [review]

**Comment by mtias** [comment]
Great! There are a few things that I think could be improved:

- Could we try flashing the block outline (same as we do when you hover the block type) when the action triggers? That should give a clearer hint of what was copied (all the block boundaries).
- It'd be nice to say "Copied {BlockName} to clipboard" if it's just one block copied.
- Does it work with "Cut"?
- We should probably add the action to the ellipsis menu but I see @ellatrix is already looking at it here: #22214.

**Comment by ellatrix** [comment]
Yeah, I added it to the block actions. We should add the shortcut there afterwards too, but that will require a follow up PR with some ClipboardButton refactoring.

**Comment by mcsf** [comment]
@ellatrix and @mtias, thanks for the reviews.

> Do you think an e2e test could be useful?

I do, but haven't taken the time to write any yet.

> * Could we try flashing the block outline (same as we do when you hover the block type) when the action triggers? That should give a clearer hint of what was copied (all the block boundaries).

Done in 99cf4427949c46e22e0a165fb95962611f8dfb63. I'm not yet sure of the implementation, but how's the UX?

> * It'd be nice to say "Copied {BlockName} to clipboard" if it's just one block copied.

Done too, and expanded to cutting actions.

> * Does it work with "Cut"?

Yes. So far it feels alright to me.

File: packages/block-editor/src/components/copy-handler/index.js [hunk 1]
```diff
@@ -1,16 +1,88 @@
...
+function useFlashBlock() {
+	const { toggleBlockHighlight } = useDispatch( 'core/block-editor' );
+	const timeouts = useRef( [] );
+	const flashBlock = useCallback(
+		( clientId ) => {
+			toggleBlockHighlight( clientId, true );
+			const timeout = setTimeout( () => {
+				toggleBlockHighlight( clientId, false );
+			}, 1000 );
+			timeouts.current.push( timeout );
```
Comment (mcsf) [inline packages/block-editor/src/components/copy-handler/index.js:24]: This feels a lot like a hack. Does it have its place here? Should the idea of flashing be a proper `block-editor` action? Should there be different primitives for this? cc @youknowriad

Comment (youknowriad) [reply packages/block-editor/src/components/copy-handler/index.js:24]: The `dispatch('core/editor').flashBlock( clientId )` action can also be written as a generator and absorbs the timeout handling to unset the reducer state. When I call `dispatch('core/editor').flashBlock( clientId )` I should expect it to handle all the logic.

Comment (mcsf) [reply packages/block-editor/src/components/copy-handler/index.js:24]: Thanks, I agree.

One last question: did you namespace it `core/editor` intentionally, as opposed to `core/block-editor`?

Comment (youknowriad) [reply packages/block-editor/src/components/copy-handler/index.js:24]: no, not intentional :) good catch.

**Comment by mtias** [comment]
> I'm not yet sure of the implementation, but how's the UX?

Made it a bit faster.

File: packages/block-editor/src/components/copy-handler/index.js (hunk 1 above)
Comment (mcsf) [reply packages/block-editor/src/components/copy-handler/index.js]: I pushed f425709c4f, but I don't really like what I did with the introduction of a `SLEEP` control. Do you suggest a different approach, @youknowriad? And if not, should I export `sleep` from `controls`, or keep it private as now?

Comment (youknowriad) [reply packages/block-editor/src/components/copy-handler/index.js]: I think the "sleep" control is totally fine. redux-saga, for instance, has a "delay" effect which is equivalent.

**Comment by mcsf** [comment]
I think this is good now with the addition of tests. Thanks for everyone's help!

File: packages/e2e-tests/specs/editor/various/__snapshots__/copy-cut-paste-whole-blocks.test.js.snap [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+<p>2</p>
+<!-- /wp:paragraph -->
+
+<!-- wp:paragraph -->
+<p>Here is a unique string so we can test copying.</p>
+<!-- /wp:paragraph -->
+
+<!-- wp:paragraph -->
+<p></p>
+<!-- /wp:paragraph -->"
```
Comment (mcsf) [inline packages/e2e-tests/specs/editor/various/__snapshots__/copy-cut-paste-whole-blocks.test.js.snap:28]: For the record, I don't like these empty paragraphs that the pasting yields, but this is tangent to the PR itself.

**Comment by desaiuditd** [comment]
I'm seeing a potential bug with this PR on version 8.2.0 as well as 8.2.1

In Embed block, copying the embed url after selecting it in the placeholder is copying the whole block instead of just url. Please refer to the screencast.



@mcsf Could this be a regression?

**Comment by mcsf** [comment]
Good catch! That definitely sounds related. Could you file an issue and let me know? Thanks!

**Comment by tmann50** [comment]
Hi, I'm getting the same problem when trying to copy or cut individual words/lines in Custom HTML blocks. The whole block is copied instead of just the selected text.

**Comment by desaiuditd** [comment]
> Good catch! That definitely sounds related. Could you file an issue and let me know? Thanks!

Here you go @mcsf #22775

On the other hand, I was thinking, can this feature be opt in?

Copying text even within custom blocks, inner blocks and other areas is trivial and important user action. So if we make it opt-in, we can avoid such bugs from creeping into users’ daily workflow. 

Power users can opt-in for this if they want to. And that way we can limit the scope of such bugs.

**Comment by mcsf** [comment]
Fix in #22793 up for review.

> On the other hand, I was thinking, can this feature be opt in?
> 
> Copying text even within custom blocks, inner blocks and other areas is trivial and important user action. So if we make it opt-in, we can avoid such bugs from creeping into users’ daily workflow.

Bugs are always frustrating, and unfortunately they will creep up sooner or later. That shouldn't however be a reason for us to fragment the user experience by selectively disabling core features. It should instead fuel our efforts to ensure quality and stability in the project. In this case, I might have tested the feature more broadly, I might have added more varied E2E tests.

In short, I think it would be a great mistake to make this opt-in.
